File: C-Sharp/FerryLegacy/PortModel.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace FerryLegacy
{
    public class PortModel
    {
        public int Id { get; set; }
        public string Name { get; set; }

        private readonly Dictionary<int, TimeSpan> _boatAvailability = new Dictionary<int, TimeSpan>();
        private readonly List<Ferry> _boats = new List<Ferry>();

        public PortModel(Port port)
        {
            Id = port.Id;
            Name = port.Name;
        }

        public void AddBoat(TimeSpan available, Ferry boat)
        {
            if (boat != null)
            {
                _boats.Add(boat);
                _boatAvailability.Add(boat.Id, available);
            }
        }

        public Ferry GetNextAvailable(TimeSpan time)
        {
            var available = _boatAvailability.FirstOrDefault(x => time >= x.Value);
            if (available.Key == 0) return null;
            _boatAvailability.Remove(available.Key);
            var boat = _boats.Single(x => x.Id == available.Key);
            _boats.Remove(boat);
            return boat;
        }
    }
}

File: java/src/main/java/ferry/booking/PortModel.java
package ferry.booking;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class PortModel {

    public int id;
    public String name;

    private final Map<Integer, Long> boatAvailability = new HashMap<>();
    private final List<Ferry> boats = new ArrayList<>();

    public PortModel(Port port) {
        id = port.id;
        name = port.name;
    }

    public void addBoat(long available, Ferry boat) {
        if (boat != null) {
            boats.add(boat);
            boatAvailability.put(boat.id, available);
        }
    }

    public Ferry getNextAvailable(long time) {
        for (Map.Entry<Integer, Long> entry : boatAvailability.entrySet()) {
            if (time >= entry.getValue()) {
                boatAvailability.remove(entry.getKey());
                for (Ferry boat : boats) {
                    if (boat.id == entry.getKey()) {
                        boats.remove(boat);
                        return boat;
                    }
                }
                return null;
            }
        }
        return null;
    }
}
